### PR TITLE
feat(remote)!: use craft-application remote build for core22

### PR DESCRIPTION
**Pull request by syu-w**
Deprecated and removed old remote build code. Use new remote-build for core22.

Core20 can also use the new remote build if set env SNAPCRAFT_REMOTE_BUILD_STRATEGY="disable-fallback"

`--build-on` no longer supported.

Fixes: #4710

### CODE REVIEW

File: snapcraft/application.py [hunk 1]
```diff
@@ -62,6 +62,25 @@
 }
 
 
+def _get_esm_error_for_base(base: str) -> None:
```
Comment (lengau) [inline snapcraft/application.py:65]: Could you do a quick pair of unit tests for this?

**Review by lengau — APPROVED** [review]
Looks good, but please do a pair of unit tests for the new function

File: snapcraft/application.py (hunk 1 above)
Comment (syu-w) [reply snapcraft/application.py:65]: Added

File: snapcraft/application.py [hunk 2]
```diff
@@ -163,11 +182,35 @@ def _get_dispatcher(self) -> craft_cli.Dispatcher:
                 yaml_data = util.safe_yaml_load(file)
             base = yaml_data.get("base")
             build_base = yaml_data.get("build-base")
+            _get_esm_error_for_base(base)
             if "core24" in (base, build_base) or build_base == "devel":
                 # We know for sure that we're handling a core24 project
                 self._known_core24 = True
             elif any(arg in ("version", "--version", "-V") for arg in sys.argv):
                 pass
+            elif "remote-build" in sys.argv and any(
```
Comment (mr-cal) [inline snapcraft/application.py:191]: Moving this out of the command means there's a mismatch now - this code in `snapcraft.application` is being tested in `tests.unit.commands.test_remote`.

Could you move either the code or the tests?

File: snapcraft/application.py [hunk 3]
```diff
@@ -163,11 +182,35 @@ def _get_dispatcher(self) -> craft_cli.Dispatcher:
...
+                        "'SNAPCRAFT_REMOTE_BUILD_STRATEGY'. "
+                        "Valid values are 'disable-fallback' and 'force-fallback'."
+                    )
+                if "core20" in (base, build_base):
+                    # Use legacy snapcraft unless explicitly forced to use craft-application
+                    if build_strategy != "disable-fallback":
+                        raise errors.ClassicFallback()
+                if "core22" in (base, build_base):
+                    # Use craft-application unless explicitly forced to use legacy snapcraft
+                    if build_strategy == "force-fallback":
```
Comment (mr-cal) [inline snapcraft/application.py:212]:
```suggestion
                # Use craft-application unless explicitly forced to use legacy snapcraft
                if "core22" in (base, build_base) and build_strategy == "force-fallback":    
```

File: snapcraft/application.py [hunk 4]
```diff
@@ -163,11 +182,35 @@ def _get_dispatcher(self) -> craft_cli.Dispatcher:
...
+                    None,
+                ):
+                    raise errors.SnapcraftError(
+                        f"Unknown value {build_strategy!r} in environment variable "
+                        "'SNAPCRAFT_REMOTE_BUILD_STRATEGY'. "
+                        "Valid values are 'disable-fallback' and 'force-fallback'."
+                    )
+                if "core20" in (base, build_base):
+                    # Use legacy snapcraft unless explicitly forced to use craft-application
+                    if build_strategy != "disable-fallback":
```
Comment (mr-cal) [inline snapcraft/application.py:208]:
```suggestion
                # Use legacy snapcraft unless explicitly forced to use craft-application
                if "core20" in (base, build_base) and build_strategy != "disable-fallback":             
```

File: snapcraft/application.py [hunk 5]
```diff
@@ -163,11 +182,35 @@ def _get_dispatcher(self) -> craft_cli.Dispatcher:
...
+                    "disable-fallback",
+                    "",
+                    None,
+                ):
+                    raise errors.SnapcraftError(
+                        f"Unknown value {build_strategy!r} in environment variable "
+                        "'SNAPCRAFT_REMOTE_BUILD_STRATEGY'. "
+                        "Valid values are 'disable-fallback' and 'force-fallback'."
+                    )
+                if "core20" in (base, build_base):
```
[the remark this answers is not in the document]
Comment (syu-w) [reply snapcraft/application.py:206]: Should this move to craft-application? Or some of our craft-app support shallow clone?

[the remark this answers is not in the document]
Comment (mr-cal) [reply snapcraft/application.py:206]: Good point.  The code was already migrated to craft-application ([source](https://github.com/canonical/craft-application/blob/53ff0ddcb5bb84984d7e0951428fca95708f31a6/craft_application/remote/git.py#L97C5-L97C36)), but it's not being called.

I think it should be called in craft-application, like Sheng suggested.  What do you think @lengau?

File: snapcraft/application.py (hunk 2 above)
Comment (syu-w) [reply snapcraft/application.py:191]: OK, I moved the tests

File: snapcraft/application.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (syu-w) [reply snapcraft/application.py]: I think I just leave the 2 checks to the craft-application

[the remark this answers is not in the document]
Comment (lengau) [reply snapcraft/application.py]: Yeah, probably should be called in craft-application, though the error message should be updated there too :-)

[the remark this answers is not in the document]
Comment (syu-w) [reply snapcraft/application.py]: OK, I will create a PR for that in craft-application

**Review by mr-cal — APPROVED** [review]
Nice work!

**Comment by syu-w** [comment]
The git repo check is in https://github.com/canonical/craft-application/pull/310

**Pull request merged**